api/v3/MessageTemplate.php

```diff
@@ -40,6 +40,27 @@
  * @throws \API_Exception
  */
 function civicrm_api3_message_template_create($params) {
+  // System Workflow Templates have a specific wodkflow_id in them but normal user end message templates don't
+  // If we have an id check to see if we are update, and need to check if original is a system workflow or not.
+  if (!empty($params['id']) && !empty($params['check_permissions'])) {
+    $details = civicrm_api3('MessageTemplate', 'getSingle', ['id' => $params['id']]);
+    if (!empty($details['workflow_id'])) {
+      if (!CRM_Core_Permission::check('edit system workflow message templates')) {
+        throw new \Civi\API\Exception\UnauthorizedException(ts('Editing or creating system workflow messages requires edit system workflow message templates permission as well as edit message templates'));
+      }
+    }
+    elseif (!CRM_Core_Permission::check('edit user-driven message templates')) {
+      throw new \Civi\API\Exception\UnauthorizedException(ts('Editing or creating user driven workflow messages requires edit user-driven message templates as well as edit message templates'));
+    }
+  }
+  elseif (!empty($params['check_permissions'])) {
+    if (!empty($params['workflow_id']) && !CRM_Core_Permission::check('edit system workflow message templates')) {
+      throw new \Civi\API\Exception\UnauthorizedException(ts('Editing or creating system workflow messages requires edit system workflow message templates permission as well as edit message templates'));
+    }
+    elseif (!CRM_Core_Permission::check('edit user-driven message templates')) {
+      throw new \Civi\API\Exception\UnauthorizedException(ts('Editing or creating user driven workflow messages requires edit user-driven message templates as well as edit message templates'));
+    }
+  }
```

Comment:
Should we move this permission check block down to BAO? Intent is to shift such checks and other manual codes down to CRUD fns and to make API more dependent on respective BAO fns.

Reply:
Exactly 😅